### ISSUE TITLE
Explore adding package signing

### ISSUE THREAD

**Issue by Eeems** [toltec-dev/build#14]
@ddvk has requested package signing:
![image](https://user-images.githubusercontent.com/1176979/93785111-f1733600-fbea-11ea-9808-ea131f92bdfb.png)
According to https://stackoverflow.com/a/39972015 it's possible. That said we need to explore the implications of doing so. There is the possibility that this would require us to handle signing all the standard entware packages.

**Comment by matteodelabre** [comment]
[Opkg LEDE](https://git.openwrt.org/project/opkg-lede.git) (OpenWrt’s fork, which Entware and us are using) supports package signing. Entware doesn’t use or support signing. There are two options that control the checking of signatures by Opkg:

* Set `option check_signature 1` in `/opt/etc/opkg.conf` to enable signature checking.
* Pass the (undocumented) `--force-signature` flag on the command line to make signature failures non-fatal.

Only the `Packages` index file is signed (its signature is located in a file called `Packages.sig`); the individual package archives are not signed. This is sufficient since the index file contains SHA-256 checksums for each package archive. Note that signature checking cannot be enabled on a per-feed basis, it’s a global setting. After downloading signatures, Opkg invokes the following script to check them:

```
/usr/sbin/opkg-key verify /path/to/Packages.sig /path/to/Packages
```

This script is tasked with checking that `/path/to/Packages.sig` is a valid signature for `/path/to/Packages`, and must exit with 0 if that’s the case. Entware does not ship such a script, so this will not work by default. There are at least two implementations of that script that we could take inspiration from:

* [This one that’s a wrapper around the `gpg` command](https://git.openwrt.org/?p=project/opkg-lede.git;a=blob;f=utils/opkg-key;h=266bb6628a9cae8096f482c4b1d78fe3c3bcb2ca;hb=HEAD)
* [This one that’s a wrapper around the `usign` command](https://github.com/Entware/Entware/blob/743d71e487d7eed48424d638f3a65aef4bfa9dc2/package/system/opkg/files/opkg-key) (see the end of this comment for more info on usign)

The script does not have to be located in `/usr/sbin/opkg-key`. This default value can be overridden in one of two ways:

* Set `option verify_program /path/to/opkg-key` in `/opt/etc/opkg.conf`.
* Pass the `--verify-program /path/to/opkg-key` flag on the command line.

***

**OpenWrt’s way:** There is a brief [page in their docs on how they do repository signatures](https://openwrt.org/docs/guide-user/security/release_signatures). They use an in-house program called [`usign`](https://git.openwrt.org/project/usign.git) to check signatures. `usign` is compatible with signatures created by BSD’s [signify](https://man.openbsd.org/OpenBSD-current/man1/signify.1) (Ed25519 algorithm).

**Comment by Eeems** [comment]
We probably want to go with `usign` to keep in line with OpenWrt. As for entware, we could host a mirror of their repository and just sign the packages in that mirror, have a script that syncs nightly or something.

**Comment by matteodelabre** [comment]
> We probably want to go with `usign` to keep in line with OpenWrt. 

Yes, `usign` seems like the best approach. It’s simpler and more targeted than GnuPG, although maybe less scrutinized from a security point of view.

> As for entware, we could host a mirror of their repository and just sign the packages in that mirror, have a script that syncs nightly or something.

Related: toltec-dev/toltec#327
